[Console] Set correct initial value in monaco editor

ElenaStoeva opened the pull request:
Closes https://github.com/elastic/kibana/issues/177332

## Summary

This PR makes the new Monaco editor for Console to have a correct initial value depending on the `initialTextValue` prop or the `load_from` url query.

### Testing

**Testing with `load_from` query parameter:**

1. Create a `config/kibana.dev.yml` file (if one doesn't exist already) and add the line: `console.dev.enableMonaco: true`
2. Start Es and Kibana
3. Navigate to `/dev_tools#/console?load_from=data:text/plain,OIUQKgBA+gzgpgQwE4GMAWAoA3gIgI4CucSAnjgFy4C2CALulAgDZMVYC+nQA`. This should prepolutate Console with the following request:
```
GET _search
{"query":{"match_all":{}}}
```

Note: The `load_from` functionality should also support loading from a URL (e.g. `app/dev_tools#/console?load_from=https://www.elastic.co/guide/en/elasticsearch/reference/current/snippets/2935.console`) but this will be addressed in a separate PR.

**Testing with `initialTextValue` prop:**
1. Create a `config/kibana.dev.yml` file (if one doesn't exist already) and add the line: `console.dev.enableMonaco: true`
2. Change line 77 in `src/plugins/console/public/application/containers/editor/editor.tsx` to `<MonacoEditor initialTextValue={<addSomeTextHere>} />`
4. Start Es and Kibana
5. Open Console and verify that it contains the text that you provided in the prop.

**Testing with `DEFAULT_INPUT_VALUE` constant:**
1. Create a `config/kibana.dev.yml` file (if one doesn't exist already) and add the line: `console.dev.enableMonaco: true`
2. Change line 77 in `src/plugins/console/public/application/containers/editor/editor.tsx` to `<MonacoEditor initialTextValue={null} />`
3. Start Es and Kibana
5. Open Console and verify that it contains the default text:
```
# Click the Variables button, above, to create your own variables.
GET ${exampleVariable1} // _search
{
  "query": {
    "${exampleVariable2}": {} // match_all
  }
}
```


<!--
### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
-->

yuliacech commented on src/plugins/console/public/application/containers/editor/monaco/monaco_editor.tsx:
```diff
@@ -6,13 +6,119 @@
+
+export const MonacoEditor = ({ initialTextValue }: EditorProps) => {
+  const {
+    services: {
```
I'm not sure the code is using all of those services, maybe we can remove the ones that are not needed?

yuliacech commented on src/plugins/console/public/application/containers/editor/monaco/monaco_editor.tsx:
```diff
@@ -6,13 +6,119 @@
+      // Normalize and encode the URL to avoid issues with spaces and other special characters.
+      const encodedUrl = new URL(url).toString();
+      if (/^https?:\/\//.test(encodedUrl)) {
+        const loadFrom: Record<string, any> = {
```
I'm not sure the variable `loadFrom` is used later in this function, I think we could remove it

ElenaStoeva replied:
Ah I missed one code block in that if-statement, nice catch! Here we parse the `load_from` param if it is an http - I added this to the test instructions in the PR.

yuliacech reviewed (commented):
Thanks a lot for working on this, @ElenaStoeva! 
I tested the functionality and it works as expected, but I think we could remove some of the code that seem to be actually not needed, could you please check my comments in the code?

ElenaStoeva commented:
Many thanks for the review @yuliacech! I addressed your comments in https://github.com/elastic/kibana/pull/178899/commits/801e00f777916911149d661c8d11c6a5a387cc94 - please let me know if you have any other suggestions.

yuliacech commented on src/plugins/console/public/application/containers/editor/monaco/monaco_editor.tsx:
```diff
@@ -6,16 +6,105 @@
+          kbnXsrfToken: false,
+        };
+
+        if (/https?:\/\/api\.github\.com/.test(url)) {
```
@ElenaStoeva I was not able to find any examples of loading a request from github, do you know how this works?

ElenaStoeva replied:
I copied this logic from the original [code editor](https://github.com/elastic/kibana/blob/ffcc1ffa3381d07b397178846c04a6b1d5fcaba2/src/plugins/console/public/application/containers/editor/legacy/console_editor/editor.tsx#L129), but I'm also not sure how exactly this works. I tracked down the PR where this was added it seems it is some legacy functionality - see https://github.com/elastic/kibana/pull/52814/files#r357049056.

yuliacech replied:
okay, thanks a lot for the info, I think we can remove that since there is no documentation about loading from github

yuliacech reviewed (approved):
Thanks a lot for working on this issue, @ElenaStoeva!  
I tested locally and all 3 options to set the value in the editor are working as expected. I'm not against merging this PR as is, but there are some issues that could be addressed either here or put into a follow up issue:
- delete the github specific headers since this use case is not documented 
- move the logic for setting the initial value, at least the `load_from` parameter into a separate file, maybe in form of a hook
- add unit tests for the 3 options of setting the initial value
- remove the usage of jquery and replace with something modern and safe, maybe the http client used in Kibana

ElenaStoeva commented:
Opened https://github.com/elastic/kibana/issues/179658 to track some follow-up work for the hook for setting the initial value.
